Notify the administrators when one of the installed licensed applications has a new version available #110

oanalavinia opened the pull request:
* add `NewExtensionVersionAvailable` event and a job scheduler to send this event every week
* update `ExtensionEventConverter` to set the custom message on the event body and `LicensingSchedulerListener` to schedule / reschedule this event as well on licensing upgrade or install
* have a `NewExtensionVersionAvailableEventDisplayer` to provide a custom template for the event
* add unit tests

![newVersionNotification](https://user-images.githubusercontent.com/22794181/168813818-9cfd153d-ca68-4362-ace6-4995194fcad5.png)

mflorea commented:
Receiving the notification that a new version of an extension is available is a plus, but the user (administrator) needs to be able to take an **action** (i.e. upgrade). We could provide links to individual extensions that the administrator could follow to perform the upgrade and / or a link to the extension updater where they can see all the available upgrades.

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/notifications/newVersion/NewExtensionVersionAvailableEventDisplayer.java:
```diff
@@ -0,0 +1,102 @@
+    {
+        XWikiContext xcontext = contextProvider.get();
+        ScriptContext scriptContext = this.scriptContextManager.getScriptContext();
+        Template customTemplate = this.templateManager.getTemplate("newVersionAvailable/newVersionAvailable.vm");
```
Is the folder ``newVersionAvailable`` really needed? Does it have multiple templates?

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/notifications/newVersion/NewExtensionVersionAvailableEventDisplayer.java:
```diff
@@ -0,0 +1,102 @@
+
+            return this.templateManager.execute(customTemplate);
+        } catch (Exception e) {
+            logger.warn("Failed to render custom template. Root cause is: [{}]", ExceptionUtils.getRootCauseMessage(e));
```
'custom template' doesn't mean much. You know which template failed to be executed.

oanalavinia commented:
I added a link to an upgrade page corresponding to each notification
![image](https://user-images.githubusercontent.com/22794181/170502128-8eea8e40-b423-4b8c-8994-f5abb6983280.png)

Also, I added an a new property to the `AutomaticUpgradesClass`, namely `newVersionNotifiedExtensions`, to keep an evidence of the already send notifications, to avoid multiplying them.
Note that I wasn't sure about using the existing `AutomaticUpgradesClass`, the name of the class might not be generic enough, but they are related.

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/LicensingConfiguration.java:
```diff
@@ -72,4 +72,17 @@
+    /**
+     * @return the list of extensions that have a new version available, that users have been notified of
+     */
+    List<String> getNewVersionNotifiedExtensions();
+
+    /**
+     * Set the list of extensions with a new version available, that users have been notified of.
+     *
+     * @param newVersionNotifiedExtensions the list of extensions with a new version available, that users have been
+     *            notified of
+     */
+    void setNewVersionNotifiedExtensions(List<String> newVersionNotifiedExtensions);
```
I don't see how this is related to configuration, which normally includes settings that the administrators can set to tweak the behavior of the application. In this case, I don't see why the administrators would set themselves the lists of extensions that have a new version available. This looks more like some **internal** state that the licensing application needs to manage in order to prevent duplicate notifications. The administrators shouldn't be aware of this.

oanalavinia replied:
What would be the best way of saving this internal state? Maybe have a separate page that holds an object with the list of extensions?

mflorea replied:
First of all, keeping all the versions for all the extensions doesn't scale (if we don't do any cleanup). The list grows indefinitely. I think we could keep only the last **known** installable version for each extension and namespace (that we have notified the administrators of). Each time this last know version changes we send a new notification. So I'm thinking we could use a class with 3 fields (namespace, extension id and version) and add objects of this type to the licensing config page, making sure we have only one object per namespace and extension id. This should be "hidden" behind a component that gives you the last known installable version and allows you to update it, in case we want to implement it differently in the future.

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/NewExtensionVersionAvailableManager.java:
```diff
@@ -77,26 +79,42 @@ public void checkLicensedExtensionsAvailableVersions()
-                namespace != null ? namespace : "root", versions.get(0));
-            this.observationManager.notify(new NewExtensionVersionAvailableEvent(), extensionId.getId(), message);
+        List<Version> installableVersions = upgradeExtensionHandler.getInstallableVersions(installedExtension.getId());
+        if (installableVersions.size() <= 0) {
```
```suggestion
        if (installableVersions.isEmpty()) {
```

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/NewExtensionVersionAvailableManager.java:
```diff
@@ -77,26 +79,42 @@ public void checkLicensedExtensionsAvailableVersions()
 
+        List<String> newVersionNotifiedExtensions = new ArrayList<>(licensingConfig.getNewVersionNotifiedExtensions());
+        String namespaceName = namespace != null ? namespace : "root";
+        // Create an identified for the extension, by consider also the version and namespace.
```
```suggestion
        // Create an identifier for the extension, by consider also the version and namespace.
```

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/NewExtensionVersionAvailableManager.java:
```diff
@@ -77,26 +79,42 @@ public void checkLicensedExtensionsAvailableVersions()
+        List<String> newVersionNotifiedExtensions = new ArrayList<>(licensingConfig.getNewVersionNotifiedExtensions());
+        String namespaceName = namespace != null ? namespace : "root";
+        // Create an identified for the extension, by consider also the version and namespace.
+        String verifiedExtensionId =
+            String.format("%s-%s-%s", extensionId.getId(), namespaceName, installableVersions.get(0));
```
Why "verified"? And it would be good to mention that the first item in ``installableVersions`` is the most recent (right?).

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/NewExtensionVersionAvailableManager.java:
```diff
@@ -77,26 +79,42 @@ public void checkLicensedExtensionsAvailableVersions()
     }
 
-    private void notifyExtensionAvailableVersions(ExtensionId extensionId, String namespace)
+    private void notifyExtensionAvailableVersion(ExtensionId extensionId, String namespace)
```
```suggestion
    private void notifyExtensionVersionAvailable(ExtensionId extensionId, String namespace)
```

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/resources/templates/newVersionAvailable.vm:
```diff
@@ -4,13 +4,29 @@
 #set ($compositeEventDate = $escapetool.xml($datetool.whenIs($event.dates.get(0))))
 
+#macro (getExtensionUpgradeURL $event)
+  ## The event body contains information about the extension id, version and targeted namespace.
+  #set ($extensionInfo = $event.getBody().split(' - '))
```
First of all, I think the message contains the extension name not id. Thus, what happens if the extension name contains ``' - '``? This is too fragile and thus error-prone. If you want to pass structured information in the message body then why not use a standard format such as JSON or XML? Or maybe you can extract this information from elsewhere.

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/NewExtensionVersionAvailableManager.java:
```diff
@@ -79,42 +79,40 @@ public void checkLicensedExtensionsAvailableVersions()
-                        new ExtensionId(extensionId.getId(), installableVersions.get(0)), namespace),
-                    extensionId.getId(), message);
+            String message = String.format("{\"extensionName\":\"%s\", \"namespace\": \"%s\", \"version\": \"%s\"}",
+                installedExtension.getName(), namespaceName, installableVersions.get(0));
```
We don't control the extension name and its version so they may contain special characters that break the JSON. Either escape the values or, better, use a Java map that you then serialize to JSON using the Jackson library that we bundled in XWiki.

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/NewVersionNotificationManager.java:
```diff
@@ -0,0 +1,157 @@
+
+    private boolean objectHasValue(BaseObject obj, String name, String value)
+    {
+        return obj.getStringValue(name) != null && obj.getStringValue(name).equals(value);
```
```suggestion
        return Objects.equals(obj.getStringValue(name), value);
```

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/test/java/com/xwiki/licensing/internal/upgrades/NewVersionNotificationManagerTest.java:
```diff
@@ -0,0 +1,187 @@
+ * @since 1.23
+ */
+@ComponentTest
+public class NewVersionNotificationManagerTest
```
```suggestion
class NewVersionNotificationManagerTest
```

mflorea commented on application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/NewExtensionVersionAvailableManager.java:
```diff
@@ -98,21 +109,25 @@ private void notifyExtensionVersionAvailable(ExtensionId extensionId, String nam
-                    namespace), extensionId.getId(), message);
+        try {
+            String namespaceName = namespace != null ? namespace : "root";
+            if (!this.newVersionNotificationManager.isNotificationAlreadySent(installedExtension.getName(),
```
Shouldn't you use the extension id instead?
